rs/ic_os/network/mac_address/src/mac_address.rs

```diff
@@ -1,129 +1,196 @@
-pub struct FormattedMacAddress(String);
-
-impl UnformattedMacAddress {
-    pub fn get(&self) -> String {
-        self.0.clone()
+impl fmt::Display for MacAddressParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            MacAddressParseError::InvalidFormat => write!(f, "Invalid MAC address format"),
+            MacAddressParseError::InvalidLength => write!(f, "Invalid MAC address length"),
+        }
     }
 }
 
-impl FormattedMacAddress {
-    pub fn get(&self) -> String {
-        self.0.clone()
+impl std::error::Error for MacAddressParseError {}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MacAddress(MacAddr6);
+
+impl MacAddress {
+    pub fn new(octets: [u8; 6]) -> Self {
+        MacAddress(MacAddr6::new(
+            octets[0], octets[1], octets[2], octets[3], octets[4], octets[5],
+        ))
     }
-}
 
-impl fmt::Display for UnformattedMacAddress {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{}", self.get())
+    pub fn octets(&self) -> [u8; 6] {
+        self.0
+            .as_bytes()
+            .try_into()
+            .expect("MAC address should always be 6 bytes")
+    }
+
+    /// Returns the unformatted MAC address string (without colons)
+    pub fn to_unformatted_string(&self) -> String {
+        self.0.to_string().replace(":", "")
+    }
+
+    /// Returns the MAC address in lowercase hexadecimal with colons
+    pub fn to_canonical(&self) -> String {
+        self.to_string().to_lowercase()
+    }
+
+    /// Generates the SLAAC IPv6 address for the given prefix.
+    pub fn calculate_slaac(&self, prefix: &str) -> Result<Ipv6Addr> {
+        let mac_octets = self.octets();
+
+        // Create the EUI-64 interface identifier
+        let mut interface_id = [0u8; 8];
+
+        // Flip the Universal/Local bit in the first octet
+        interface_id[0] = mac_octets[0] ^ 0x02;
+        interface_id[1] = mac_octets[1];
+        interface_id[2] = mac_octets[2];
+        interface_id[3] = 0xff;
+        interface_id[4] = 0xfe;
+        interface_id[5] = mac_octets[3];
+        interface_id[6] = mac_octets[4];
+        interface_id[7] = mac_octets[5];
+
+        // Prepare the prefix by appending '::' if necessary
+        let full_prefix = if prefix.contains("::") || prefix.ends_with(':') {
+            prefix.to_string()
+        } else {
+            format!("{}::", prefix)
+        };
+
+        // Parse the prefix into an Ipv6Addr
+        let prefix_addr = full_prefix
+            .parse::<Ipv6Addr>()
+            .map_err(|_| anyhow!("Invalid IPv6 prefix: {}", prefix))?;
+
+        // Extract the network prefix (first 64 bits)
+        let mut addr_octets = prefix_addr.octets();
+
+        // Combine the prefix with the interface identifier
+        addr_octets[8..].copy_from_slice(&interface_id);
+
+        // Construct the full IPv6 address
+        Ok(Ipv6Addr::from(addr_octets))
```

Comment:
This could exist separate from mac_address. It could live in deterministic_ips still and you would pass in a MacAddress object and prefix.